jans-cedarling/cedarling/src/common/policy_store.rs

```diff
@@ -5,155 +5,430 @@
+/// version, e.g. `v1.0.1`.
+fn check_cedar_version<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let version: String = String::deserialize(deserializer)?;
+
+    // Check for "v" prefix
+    let stripped_version = if version.starts_with('v') {
+        &version[1..] // Remove the 'v' prefix
+    } else {
+        &version
+    };
+
+    // Split the version by '.'
+    let parts: Vec<&str> = stripped_version.split('.').collect();
+
+    // Check that we have exactly three parts (major, minor, patch)
+    if parts.len() != 3 {
+        return Err(serde::de::Error::custom(format!(
+            "invalid version format: '{}', expected 'major.minor.patch'",
+            version
+        )));
+    }
+
+    // Check that each part is a valid number
+    for part in &parts {
+        if part.parse::<u32>().is_err() {
+            return Err(serde::de::Error::custom(format!(
+                "invalid version part: '{}', expected a number",
+                part
+            )));
+        }
+    }
+
+    Ok(version)
+}
+
+/// Enum representing various error messages that can occur while parsing policy sets.
+///
+/// This enum is used to provide detailed error information during the deserialization
+/// of policy sets within the `PolicyStore`.
+#[derive(Debug, thiserror::Error)]
+enum ParsePolicySetMessage {
+    /// Indicates failure to decode policy content as base64.
+    #[error("unable to decode policy_content as base64")]
+    Base64,
+
+    /// Indicates failure to decode policy content to a UTF-8 string.
+    #[error("unable to decode policy_content to utf8 string")]
+    String,
+
+    /// Indicates failure to decode policy content from a human-readable format.
+    #[error("unable to decode policy_content from human readable format")]
+    HumanReadable,
+
+    /// Indicates failure to collect policies into a policy set.
+    #[error("could not collect policy store's to policy set")]
+    CreatePolicySet,
+}
+
+/// Represents a raw policy entry from the `PolicyStore`.
+///
+/// This is a helper struct used internally for parsing base64-encoded policies.
+#[derive(Debug, serde::Deserialize)]
+struct RawPolicy {
+    /// Base64-encoded content of the policy.
+    pub policy_content: String,
+}
+
+/// Custom deserializer for converting base64-encoded policies into a `PolicySet`.
+///
+/// This function is used to deserialize the `policies` field in `PolicyStore`.
+pub fn parse_cedar_policy<'de, D>(deserializer: D) -> Result<cedar_policy::PolicySet, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let policies = <HashMap<String, RawPolicy> as serde::Deserialize>::deserialize(deserializer)?;
+
+    let policy_vec = policies
+        .into_iter()
+        .map(|(id, policy_raw)| {
+            let policy = parse_single_policy::<D>(&id, policy_raw).map_err(|err| {
+                serde::de::Error::custom(format!(
```

Comment:
It can be simplified to

```suggestion
    let policy_vec = policies
        .into_iter()
        .map(|(id, policy_raw)| {
            parse_single_policy::<D>(&id, policy_raw).map_err(|err| {
                serde::de::Error::custom(format!(
                    "unable to decode policy with id: {id}, error: {err}"
                ))
            })
        })
        .collect::<Result<Vec<cedar_policy::Policy>, _>>()?;
```

Reply:
Here you go [e36f963](https://github.com/JanssenProject/jans/pull/9910/commits/e36f9639e2d4ed1d9594fd6730487b30731b8ac2)